Volodin Evgeniy. Lab №3 Option №1.

evldn opened the pull request:
In this lab, a pass was implemented that replaces the argument-dependent mul and add instructions with one - muladd.
**The source code:**
``` C++
#include <immintrin.h>

__m128d simple_test(__m128d a, __m128d b, __m128d c, __m128d d) {
	__m128d temp = a * b;
	d = temp + c;
	return d;
}
```
**LLVM MIR code before applying the pass:**
``` MIR
body:             |
  bb.0.entry:
    liveins: $xmm0, $xmm1, $xmm2
  
    renamable $xmm0 = nofpexcept MULPDrr killed renamable $xmm0, killed renamable $xmm1, implicit $mxcsr
    renamable $xmm0 = nofpexcept ADDPDrr killed renamable $xmm0, killed renamable $xmm2, implicit $mxcsr
    RET64 $xmm0
```
**LLVM MIR code after applying the pass:**
``` MIR
body:             |
  bb.0.entry:
    liveins: $xmm0, $xmm1, $xmm2
  
    $xmm0 = VFMADD213PDr $xmm0, $xmm1, $xmm2, implicit $mxcsr
    RET64 $xmm0
```

evldn committed:
Lab3 - commit 1.

evldn committed:
Lab3 - commit 2 (reload).

evldn committed:
Lab3 - commit 3 (fix test).

evldn committed:
Lab3 - commit 4 (fix test).

evldn committed:
Lab3 - commit 5 (fix test).

evldn committed:
Lab3 - commit 6 (last fix test, i hope).

evldn committed:
Lab3 - commit 7 (reload ubuntu and fix clang-format).

m-ly4 commented on llvm/lib/Target/X86/CMakeLists.txt:
```diff
@@ -23,6 +23,7 @@ tablegen(LLVM X86GenFoldTables.inc -gen-x86-fold-tables -asmwriternum=1)
 add_public_tablegen_target(X86CommonTableGen)
 
 set(sources
+  X86VolodinEMulAddPass.cpp
```
Please rewrite it as shared lib plugin.

m-ly4 commented on llvm/lib/Target/X86/X86VolodinEMulAddPass.cpp:
```diff
@@ -0,0 +1,95 @@
+bool findInstruction(std::vector<MachineInstr *> &MIvector,
+                     MachineInstr *instruction) {
+  auto result{std::find(begin(MIvector), end(MIvector), instruction)};
+  if (result == end(MIvector))
+    return false;
+  else
+    return true;
```
```suggestion
  return (result != end(MIvector));
```

m-ly4 commented on llvm/lib/Target/X86/X86VolodinEMulAddPass.cpp:
```diff
@@ -0,0 +1,95 @@
+
+  for (MachineBasicBlock &MBB : MF) {
+    for (auto iterator = MBB.begin(); iterator != MBB.end(); ++iterator) {
+      if (iterator->getOpcode() == X86::MULPDrr) {
```
```suggestion
      if (iterator->getOpcode() != X86::MULPDrr) {
        continue;
      }
```

m-ly4 commented on llvm/test/lab3/test_muladd.mir:
```diff
@@ -0,0 +1,387 @@
+# 	e = a + c;
+# 	return d - e;
+# }
+
```
Please add also:
```
# __m128d another_test(__m128d a, __m128d b, __m128d c, __m128d d) {
# 	__m128d t = a * b;
# 	__m128d t2 = c * d;
# 	t2 = t + t2;
# 	t = t + c;
# 	return t * t2;
# }
```

evldn committed:
Lab3 - commit 8 (add test and fix pass).

evldn commented:
1. I rewrote pass as shared lib plugin.
2. I added the test you suggested, which was performed incorrectly on the old version of the pass. I fixed this by adding additional checks to the pass that track the mutability of operands between the mul and add instructions.

evldn committed:
Lab3 - commit 9 (fix X86.h - delete old function).
